rust/xain-coordinator/src/training.rs

```diff
@@ -0,0 +1,337 @@
+use std::{
+    any::Any,
+    collections::{BTreeMap, BTreeSet},
+    fmt,
+    time::Duration,
+};
+
+use crate::{Model, ModelDim};
+
+macro_rules! err {
+    ($($tt:tt)*) => {
+        ProtocolError { msg: format!($($tt)*) }
+    }
+}
+
+macro_rules! bail {
+    ($($tt:tt)*) => {
+        return Err(err!($($tt)*))
+    }
+}
+
+#[derive(Debug)]
+pub struct ProtocolError {
+    msg: String,
+}
+
+impl fmt::Display for ProtocolError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self.msg, f)
+    }
+}
+
+impl std::error::Error for ProtocolError {}
+
+type ParticipantId = u32;
+
+#[derive(Debug)]
+pub struct FromParticipant<T> {
+    pub from: ParticipantId,
+    pub payload: T,
+}
+
+#[derive(Debug)]
+pub enum InMessage {
+    Joined(FromParticipant<()>),
+    Update(FromParticipant<Model>),
+    RoundTimeout,
+}
+
+impl InMessage {
+    pub fn new_joined(from: ParticipantId) -> InMessage {
+        InMessage::Joined(FromParticipant { from, payload: () })
+    }
+
+    pub fn new_update(from: ParticipantId, model: Model) -> InMessage {
+        InMessage::Update(FromParticipant { from, payload: model })
+    }
+
+    pub fn new_round_timeout() -> InMessage {
+        InMessage::RoundTimeout
+    }
+}
+
+#[derive(Debug)]
+pub struct OutMessage {
+    pub to: ParticipantId,
+    pub kind: OutMessageKind,
+}
+
+#[derive(Debug)]
+pub enum OutMessageKind {
+    Update { model: Model },
+    Finished { model: Model },
+}
+
+pub trait IO {
+    fn send(&mut self, msg: OutMessage);
+    fn schedule_timeout(&mut self, duration: Duration) -> TimeoutToken;
+}
+
+/// Dropping `TimeoutToken` cancels the associated timeout.
+pub struct TimeoutToken {
+    pub on_cancel: Box<dyn Any>,
+}
+
+pub struct TrainingParams {
+    model_dim: ModelDim,
+    initial_model: Model,
+    n_participants: u32,
+    n_rounds: u32,
+    round_timeout: Duration,
+}
+
+pub struct Training {
+    params: TrainingParams,
+    state: TrainingState,
+}
+
+enum TrainingState {
+    Joining { participants: BTreeSet<ParticipantId> },
+    Round(RoundState),
+    Finished,
+}
+
+struct RoundState {
+    participants: BTreeSet<ParticipantId>,
+    round_results: BTreeMap<ParticipantId, Model>,
+    model: Model,
+    round: u32,
+    timeout_token: Option<TimeoutToken>,
+}
+
+impl Training {
+    pub fn new(params: TrainingParams) -> Training {
+        Training { params, state: TrainingState::Joining { participants: BTreeSet::new() } }
+    }
+
+    pub fn on_message(&mut self, message: InMessage, io: &mut dyn IO) -> Result<(), ProtocolError> {
+        log::info!("on_message({:?})", message);
+        match (&mut self.state, &message) {
+            (TrainingState::Joining { participants }, InMessage::Joined(joined)) => {
+                let is_new = participants.insert(joined.from);
```

Comment:
I'm unsure of this case: joining twice might happen when the first join only half succeeds (connection breaking suddenly). I think joining twice should be valid.

Maybe, joining should carry a _participant state_.

Reply:
Good question! I think behavior in case of duplicate messages should be explicitly designed. I think we can assume that transport layer guarantees that messages are delivered at most once, so we only need to protect from cases where the client stops and restarts. Can we push this responsibility to the transport layer? That is, make the transport layer tag each message beetwen join/leave with a session cookie, and enforce the invariant that sessions are disjoint?